Budget report generation error

If I create a budget report containing a table with a row that contains formulas, budgeteer fails with a parsing error.

It should instead dump the furmulas identically to the output just like plain content nodes.

E.g., if I replace cell xxx with the text 'Mars', every budget row starts “from Mars.” If I, on the other hand, replace it the formula that does more complicate calculations e.g. by calling the Excel function XVERWEIS (internal name xlookup), budgeteer displays an internal error, stack trace:

```java
Caused by: org.apache.poi.ss.formula.eval.NotImplementedException: Error evaluating cell Gesamtübersicht!E4
        at org.apache.poi.ss.formula.WorkbookEvaluator.addExceptionInfo(WorkbookEvaluator.java:347)
        at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateAny(WorkbookEvaluator.java:288)
        at org.apache.poi.ss.formula.WorkbookEvaluator.evaluate(WorkbookEvaluator.java:220)
        at org.apache.poi.xssf.usermodel.BaseXSSFFormulaEvaluator.evaluateFormulaCellValue(BaseXSSFFormulaEvaluator.java:64)
        at org.apache.poi.ss.formula.BaseFormulaEvaluator.evaluateFormulaCellEnum(BaseFormulaEvaluator.java:192)
        at org.apache.poi.ss.formula.BaseFormulaEvaluator.evaluateAllFormulaCells(BaseFormulaEvaluator.java:267)
        at org.apache.poi.ss.formula.BaseFormulaEvaluator.evaluateAllFormulaCells(BaseFormulaEvaluator.java:258)
        at org.apache.poi.xssf.usermodel.XSSFFormulaEvaluator.evaluateAllFormulaCells(XSSFFormulaEvaluator.java:72)
        at org.wickedsource.budgeteer.service.budget.report.BudgetReportService.createReportFile(BudgetReportService.java:75)
        at WICKET_org.wickedsource.budgeteer.service.budget.report.BudgetReportService$$FastClassByCGLIB$$f3a30fe1.invoke(<generated>)
        at net.sf.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
        at org.apache.wicket.proxy.LazyInitProxyFactory$AbstractCGLibInterceptor.intercept(LazyInitProxyFactory.java:350)
        at WICKET_org.wickedsource.budgeteer.service.budget.report.BudgetReportService$$EnhancerByCGLIB$$18923a9e.createReportFile(<generated>)
        at org.wickedsource.budgeteer.web.pages.budgets.overview.report.BudgetReportFileModel.load(BudgetReportFileModel.java:42)
        at org.wickedsource.budgeteer.web.pages.budgets.overview.report.BudgetReportFileModel.load(BudgetReportFileModel.java:14)
        at org.apache.wicket.model.LoadableDetachableModel.getObject(LoadableDetachableModel.java:135)
        at org.wickedsource.budgeteer.web.pages.budgets.overview.report.form.BudgetReportForm.onSubmit(BudgetReportForm.java:114)
        at org.apache.wicket.markup.html.form.Form.delegateSubmit(Form.java:1315)
        at org.apache.wicket.markup.html.form.Form.process(Form.java:976)
        at org.apache.wicket.markup.html.form.Form.onFormSubmitted(Form.java:797)
        at org.apache.wicket.markup.html.form.Form.onFormSubmitted(Form.java:708)
        ... 74 common frames omitted
Caused by: org.apache.poi.ss.formula.eval.NotImplementedFunctionException: _xlfn.XLOOKUP
        at org.apache.poi.ss.formula.UserDefinedFunction.evaluate(UserDefinedFunction.java:56)
        at org.apache.poi.ss.formula.OperationEvaluatorFactory.evaluate(OperationEvaluatorFactory.java:129)
        at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateFormula(WorkbookEvaluator.java:514)
        at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateAny(WorkbookEvaluator.java:278)
        ... 93 common frames omitted
```

I suggest that the replacement parser first tests the cell type and skips everything this is not of type text.

Find attached examples based on the “official budget report example:”
* [Mars Example_budget_template.xlsx](https://github.com/adessoAG/budgeteer/files/4993477/Mars.Example_budget_template.xlsx)
* [XVerweis Example_budget_template.xlsx](https://github.com/adessoAG/budgeteer/files/4993476/XVerweis.Example_budget_template.xlsx)

--- a/budgeteer-web-interface/src/main/java/org/wickedsource/budgeteer/service/budget/report/BudgetReportService.java
+++ b/budgeteer-web-interface/src/main/java/org/wickedsource/budgeteer/service/budget/report/BudgetReportService.java
@@ -72,7 +72,6 @@ public class BudgetReportService {
 		writeSummary(wb.getSheetAt(0), overallSummary);
 		writeSummary(wb.getSheetAt(1), monthlySummary);
 
-		XSSFFormulaEvaluator.evaluateAllFormulaCells(wb);
 		return createOutputFile(wb);
 	}
 

--- a/budgeteer-web-interface/src/main/java/org/wickedsource/budgeteer/service/budget/report/BudgetReportService.java
+++ b/budgeteer-web-interface/src/main/java/org/wickedsource/budgeteer/service/budget/report/BudgetReportService.java
@@ -1,5 +1,7 @@
 package org.wickedsource.budgeteer.service.budget.report;
 
+import org.apache.poi.ss.formula.eval.NotImplementedException;
+import org.apache.poi.ss.formula.eval.NotImplementedFunctionException;
 import org.apache.poi.ss.usermodel.Sheet;
 import org.apache.poi.xssf.usermodel.XSSFFormulaEvaluator;
 import org.apache.poi.xssf.usermodel.XSSFSheet;
@@ -72,6 +74,11 @@ public class BudgetReportService {
 		writeSummary(wb.getSheetAt(0), overallSummary);
 		writeSummary(wb.getSheetAt(1), monthlySummary);
 
+		try {
+			XSSFFormulaEvaluator.evaluateAllFormulaCells(wb);
+		} catch (NotImplementedException e) {
+			wb.setForceFormulaRecalculation(true);
+		}
 		return createOutputFile(wb);
 	}
 

